### PR TITLE
Sudomain redirect

### PR DESCRIPTION
**Pull request by RangerMauve**
- Adds new flags
- Can now serve subdomains just like IPFS Gateways
- Can redirect paths to subdomain

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,6 +44,18 @@ You can load data from the gateway by specifying a Hyperdrive key and a path to
 The specific HTTP verbs and headers that are supported can be found in [hypercore-fetch](https://github.com/RangerMauve/hypercore-fetch).
 Basically you can replace the `hyper://` in a URL with `http://localhost:4973/hyper/` and it'll work.
 
+## Subdomain / Subdomain-Redirect
+
+When serving contents on a gateway, some sites may be using absolute URLs (e.g. `/scrtipt.js`) which can get messed up when the data is being served from `/hyper/key_here/`.
```
Comment (jackyzha0) [inline README.md]:
```suggestion
When serving contents on a gateway, some sites may be using absolute URLs (e.g. `/script.js`) which can be messed up when the data is being served from a subdomain (e.g. `http://some-gateway.com/hyper/key_here/`).
```

Is this a better description?

File: src/index.js [hunk 2]
```diff
@@ -44,14 +48,56 @@ export async function create ({
...
+
+      if (subdomain || subdomainRedirect) {
+        const { host } = headers
+        if (host) {
+          const segments = host.split('.')
+          const [subdomain] = segments
+          if (segments.length > 2) {
+            isSubdomained = true
+
+            if (subdomain.length === 52) {
```
Comment (jackyzha0) [inline src/index.js]: perhaps pull number out instead of magic constant

File: src/index.js [hunk 3]
```diff
@@ -44,14 +48,56 @@ export async function create ({
...
+          return
+        }
+
+        finalURL = HYPER_PREFIX + url.slice('/hyper/'.length)
+
+        if (subdomainRedirect && !isSubdomained) {
+        // Take the host from the URL and redirect to the subdomain
+          const host = headers.host
+          const [hyperHost, ...pathSegments] = finalURL.slice(HYPER_PREFIX.length).split('/')
+          const subdomainName = hyperHost.replaceAll('-', '--').replaceAll('.', '-')
```
Comment (jackyzha0) [inline src/index.js]: maybe we should have a function for escaping and unescaping domains instead of repeating the logic

File: src/index.js [hunk 4]
```diff
@@ -44,14 +48,56 @@ export async function create ({
...
+          const redirectURL = `//${subdomainName}.${host}${pathSegments.join('/') || '/'}`
```
Comment (jackyzha0) [inline src/index.js]: is the redirect URL supposed to just start with `//`

Comment (RangerMauve) [reply src/index.js]: Yeah, `//` in a URL means it's protocol agnostic, so it'll use either HTTP or HTTPS depending on what URL it's redirecting from. I still need to double check that it actually works though. :P

Comment (jackyzha0) [reply src/index.js]: Ok I just checked, I think it's :// not //

**Comment by jackyzha0** [comment]
Bump! Did this ever get updated?

**Comment by YurkoWasHere** [comment]
Pulled compiled and running on Hypha's server
Working well

File: src/index.js (hunk 4 above)
Comment (RangerMauve) [reply src/index.js]: Perfect TY

**Comment by RangerMauve** [comment]
Out in hyper-gateway@3.2.1 TY for the review @jackyzha0 !

**Pull request merged**